src/lib/shell/commands/Ota.cpp

```diff
@@ -0,0 +1,268 @@
+    }
+}
+
+void OnQueryImageFailure(void * /* context */, EmberAfStatus status)
+{
+    ChipLogError(SoftwareUpdate, "QueryImage failed: %" PRIu16, static_cast<uint16_t>(status));
+}
+
+void OnQueryImageConnection(void * /* context */, DeviceProxy * deviceProxy)
+{
+    // Initialize cluster object
+    Controller::OtaSoftwareUpdateProviderCluster cluster;
+    CHIP_ERROR error = cluster.Associate(deviceProxy, kProviderEndpointId);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "Associate() failed: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    // Send QueryImage command
+    uint16_t vendorId        = 0;
+    uint16_t productId       = 0;
+    uint16_t hardwareVersion = 0;
+    uint16_t softwareVersion = 0;
+
+    DeviceLayer::ConfigurationMgr().GetVendorId(vendorId);
+    DeviceLayer::ConfigurationMgr().GetProductId(productId);
+    DeviceLayer::ConfigurationMgr().GetProductRevision(hardwareVersion);
+    DeviceLayer::ConfigurationMgr().GetFirmwareRevision(softwareVersion);
+
+    QueryImage::Type request;
+    request.vendorId           = static_cast<VendorId>(vendorId);
+    request.productId          = productId;
+    request.softwareVersion    = softwareVersion;
+    request.protocolsSupported = kRequestorProtocols;
+    request.hardwareVersion.SetValue(hardwareVersion);
+    request.location.Emplace(kRequestorLocation);
+    request.requestorCanConsent.SetValue(kRequestorCanConsent);
+    error = cluster.InvokeCommand(request, /* context */ nullptr, OnQueryImageResponse, OnQueryImageFailure);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(SoftwareUpdate, "QueryImage() failed: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+}
+
+void OnConnectionFailure(void * /* context */, NodeId nodeId, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "Connection failed: %" CHIP_ERROR_FORMAT, error.Format());
+}
+
+void QueryImageAsyncHandler(intptr_t)
+{
+    static Callback::Callback<OnDeviceConnected> successCallback(OnQueryImageConnection, nullptr);
+    static Callback::Callback<OnDeviceConnectionFailure> failureCallback(OnConnectionFailure, nullptr);
+
+    OperationalDeviceProxy * deviceProxy = Server::GetInstance().GetOperationalDeviceProxy();
+    VerifyOrReturn(deviceProxy != nullptr);
+
+    deviceProxy->UpdateDeviceData(sProviderAddress, deviceProxy->GetMRPIdleInterval(), deviceProxy->GetMRPActiveInterval());
+    deviceProxy->Connect(&successCallback, &failureCallback);
+}
+
+CHIP_ERROR QueryImageHandler(int argc, char ** argv)
+{
+    VerifyOrReturnError(argc == 4, CHIP_ERROR_INVALID_ARGUMENT);
+
+    const FabricIndex providerFabricIndex = static_cast<FabricIndex>(strtoul(argv[0], nullptr, 10));
+    const NodeId providerNodeId           = static_cast<NodeId>(strtoull(argv[1], nullptr, 10));
+    const char * ipAddressStr             = argv[2];
+    const uint16_t udpPort                = static_cast<uint16_t>(strtoul(argv[3], nullptr, 10));
```

Comment:
There should be no need to do this manually. It should be done using operational discovery. Otherwise we keep adding more and more code that is temporary and will get cut'n'pasted across many examples. This is already happening with OTA where both Nordic and Espressif have made various examples of OTA requestor client by cut'n'pasting a lot of already example-only code.

Reply:
Right, that's why I want to create a shell command so that other platform vendors don't have to create their OTA requestor clients. When it comes to specifying the address, currently there's no clear API for initiating a connection without knowing the address and OTA feature must be tested in TE7.5, so we don't have much time left. This may change soon as I've seen a PR which changes the API, but it's not yet on master.

Reply:
This is clashing, to a degree with https://github.com/project-chip/connectedhomeip/pull/12013